ocrd/validator/page_validator.py

```diff
@@ -0,0 +1,174 @@
+import re
+
+from ocrd.utils import getLogger
+from ocrd.model.ocrd_page import from_file, parse
+
+from ocrd.model.ocrd_page_generateds import (
+    PcGtsType,
+    PageType,
+    TextEquivType,
+    TextRegionType,
+    TextLineType,
+    WordType,
+    GlyphType,
+)
+from .report import ValidationReport
+
+log = getLogger('ocrd.page_validator')
+
+_HIERARCHY = [
+    (PageType,       'Page',       'get_TextRegion', None), # pylint: disable=bad-whitespace
+    (TextRegionType, 'TextRegion', 'get_TextLine',   '\n'), # pylint: disable=bad-whitespace
+    (TextLineType,   'TextLine',   'get_Word',       ' '),  # pylint: disable=bad-whitespace
+    (WordType,       'Word',       'get_Glyph',      ''),   # pylint: disable=bad-whitespace
+    (GlyphType,      'Glyph',      None,             None), # pylint: disable=bad-whitespace
+]
+
+class ConsistencyError(Exception):
+
+    def __init__(self, tag, ID, actual, expected):
+        self.tag = tag
+        self.ID = ID
+        self.actual = actual
+        self.expected = expected
+        super(ConsistencyError, self).__init__("INCONSISTENCY in %s ID '%s': text results '%s' != concatenated '%s'" % (tag, ID, actual, expected))
+
+class PageValidator(object):
+
+    @staticmethod
+    def validate_filename(filename, **kwargs):
+        """
+        Validates a PAGE file for consistency by filename
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        validator = PageValidator(parse(filename, silence=True), **kwargs)
+        return validator.validate()
+
+    @staticmethod
+    def validate_ocrd_page(page, **kwargs):
+        """
+        Validates an OcrdPage for consistency
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        validator = PageValidator(page, **kwargs)
+        return validator.validate()
+
+
+    @staticmethod
+    def validate_ocrd_file(ocrd_file, **kwargs):
+        """
+        Validates an OcrdFile representing a PAGE doc for consistency
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        validator = PageValidator(from_file(ocrd_file), **kwargs)
+        return validator.validate()
+
+    def __init__(self, page, strictness='strict'):
+        """
+        Arguments:
+            strictness (string): 'strict', 'fix' or 'off'
+        """
+        self.report = ValidationReport()
+        self.page = page
+        self.strictness = strictness
+
+    def validate(self):
+        """
+        Do the validation / fixing.
+
+        Returns:
+            report (:class:`ValidationReport`) Report on the validity
+        """
+        if self.strictness == 'off':
+            return self.report
+        self.find_or_fix_inconsistencies(self.page, strictness=self.strictness)
+        return self.report
+
+    def concatenate_index_1_text(self, node, concatenate_with):
```

Comment:
Not sure whether it makes much sense in real life but `index` (c|sh)ould be parameter.

Reply:
I don't think it would ever make sense to have a "concatenate by index 2" method. But it **would** make sense to have a "concatenate most confident" method. So maybe have the signature

```
def concatenate_text(self, node, strategy='index1', concatenate_with=' '):
  if strategy not in ('index1', 'conf'):
    raise Exception("No such strategy %s" % strategy)
```

Reply:
YES!

Reply:
Now checked in constructor